### ISSUE TITLE
extremely slow realloc performance

### ISSUE THREAD

**Issue by shadeware** [microsoft/mimalloc#153]
I was trying to benchmark mimalloc on private project, along with jemalloc, tcmalloc and libc-malloc. Last 3 implementations passed tests flawlessly in under a second, however mimalloc hanged on serialization of random struct, spending over 20000 seconds (!) on it. 

Upon closer inspection it turned out that the serialization itself is implemented in a weird way: it calls realloc for every new element, even if it's just a single integer field. That was probably done for cheap implementation of vector in C, and I've seen several snippets like this in the past.

If slowdown was 2-10 times then I'd ignore this as an unlikely and uninteresting case; however it has exponential complexity and can easily take hours or even years. I can't guarantee that one of my third-party libraries doesn't use this pattern and therefore it's a blocker for me.

Here's the minimal C++ code that shows the problem:

```cpp
#include <string>
#include <cmemory>

int main(int argc, char ** argv)
{
  int max_size = std::stoi(argv[1]);

  long long *p = (long long*) malloc(8);
  for (int i = 2; i < max_size; ++i)
  {
    p = (long long*) realloc(p, i * 8);
    p[i - 1] = i;
  }
}
```

Running this code shows that after a certain threshold time starts to increase exponentially:

|argument | time elapsed|
|---|---|
250000 | 0.0 sec
260000 | 0.0 sec
270000 | 3.1 sec
280000 | 7.9 sec
290000 | 13.2 sec
300000 | 19.3 sec
310000 | 25.1 sec
400000 | 86.5 sec
500000 | 210.9 sec

In comparison, other 3 malloc implementations performed in 0.0 sec at 500000 elements.

If you don't want to change your algorithm, you probably can still detect such pattern and fallback to another implementation.

**Comment by degski** [comment]
Technically that code (in C++) is UB, so there's no way to reason about it.

**Comment by daanx** [comment]
Thanks for the report! Of course, `realloc` can at worst be O(n) if it cannot expand in place, and therefore the given program is also at worst O(n^2) -- it happens to work for most allocators but nothing is guaranteed. Nevertheless, the main point of `mimalloc` is that is supposed to consistently do well so thanks for spotting this case! Turns out that once we get to "huge" objects (over 2MiB in mimalloc) that mimalloc would use exact sizes instead of using 12.5% incremented alignment to give headroom for potential in-place expansion. This is fixed in the latest dev branch ( d278c26) 
Hope this fixes your issues :-)

**Comment by daanx** [comment]
Ah yes, that's right.  I was a bit imprecise in my comment, but the actual commit tries to do this: it picks at the alignment at 1/8th of the requested size, but aligned to the OS page size (4KiB), and at most up to 4MiB.  Still, perhaps we should use the allocation granularity (64KiB on windows) instead of the page size... or indeed even just pick a fixed 256KiB (= 1/8 * 2MiB). I will take another look.
